File: leetcode/1209-remove-all-adjacent-duplicates-in-string-ii/Solution.java
import java.util.ArrayList;
import java.util.List;

class Solution {
    public String removeDuplicates(String s, int k) {
        List<int[]> stack = new ArrayList<>(); // each entry: [char, count]
        for (char ch : s.toCharArray()) {
            if (!stack.isEmpty() && stack.get(stack.size() - 1)[0] == ch) {
                stack.get(stack.size() - 1)[1]++;
            } else {
                stack.add(new int[]{ch, 1});
            }
            if (stack.get(stack.size() - 1)[1] == k) {
                stack.remove(stack.size() - 1);
            }
        }

        StringBuilder sb = new StringBuilder();
        for (int[] entry : stack) {
            for (int i = 0; i < entry[1]; i++) {
                sb.append((char) entry[0]);
            }
        }
        return sb.toString();
    }
}

File: leetcode/1209-remove-all-adjacent-duplicates-in-string-ii/solution.cpp
#include <string>
#include <vector>
using namespace std;

class Solution {
public:
    string removeDuplicates(string s, int k) {
        vector<pair<char, int>> stack;
        for (char ch : s) {
            if (!stack.empty() && stack.back().first == ch) {
                stack.back().second++;
            } else {
                stack.push_back({ch, 1});
            }
            if (stack.back().second == k) {
                stack.pop_back();
            }
        }

        string result;
        for (auto& [ch, count] : stack) {
            result.append(count, ch);
        }
        return result;
    }
};
